### PR TITLE
update notifier for 2019

### PR DESCRIPTION
**Pull request by ariovistus**

### CODE REVIEW

**Comment by ariovistus** [comment]
probably depends on resource update

File: wpilib/wpilib/notifier.py [hunk 1]
```diff
@@ -38,15 +38,17 @@ def __init__(self, run: callable) -> None:
...
         self._thread.start()
         
         # python-specific
         Resource._add_global_resource(self)
     
-    # python-specific: this is called finalize in Java
-    def free(self) -> None:
+    def close(self) -> None:
         handle, self._notifier = self._notifier, None
+        if handle == 0:
```
Comment (virtuald) [inline wpilib/wpilib/notifier.py]: This should be `if not handle`?

File: wpilib/wpilib/resource.py [hunk 2]
```diff
@@ -40,15 +40,15 @@ def _reset():
                 if ref is None:
                     continue
                 obj = ref()
-                if obj is not None and hasattr(obj, 'free'):
-                    obj.free()
+                if obj is not None and hasattr(obj, 'close'):
```
Comment (virtuald) [inline wpilib/wpilib/resource.py]: This needs to allow free and close.

Comment (ariovistus) [reply wpilib/wpilib/resource.py]: if both are defined, can close handle being called twice?

Comment (virtuald) [reply wpilib/wpilib/resource.py]: It should. But, I think just call one or the other, preferring close.

File: wpilib/wpilib/resource.py [hunk 3]
```diff
@@ -40,14 +40,18 @@ def _reset():
...
                     obj.free()
             
             resource.numAllocated = [None]*len(resource.numAllocated)
         
         for ref in Resource._global_resources:
             obj = ref()
-            if obj is not None and hasattr(obj, 'free'):
+            if obj is not None and hasattr(obj, 'close'):
+                obj.close()
+            elif obj is not None and hasattr(obj, 'free'):
```
Comment (auscompgeek) [inline wpilib/wpilib/resource.py]: Maybe refactor these to not repeat `obj is not None` as well.

**Comment by auscompgeek** [comment]
Needs updating again and verification.

**Comment by ariovistus** [comment]
upstream added a setUncaughtExceptionHandler here: https://github.com/wpilibsuite/allwpilib/blob/master/wpilibj/src/main/java/edu/wpi/first/wpilibj/Notifier.java#L124

I'm assuming we're going to do what dustin wants: https://github.com/wpilibsuite/allwpilib/pull/926
?

**Comment by auscompgeek** [comment]
Yeah, I don't see catching uncaught exceptions being particularly useful in Notifier for us. Threads should exit cleanly and die if the main thread dies for us anyway.

File: wpilib/wpilib/notifier.py [hunk 4]
```diff
@@ -77,6 +92,8 @@ def _run(self) -> None:
                 if self._periodic:
                     self._expirationTime += self._period
                     self._updateAlarm()
+                else:
+                    self._updateAlarm(-1)
```
Comment (auscompgeek) [inline wpilib/wpilib/notifier.py]: You should add in the comment from upstream here explaining what this is for.

Comment (ariovistus) [reply wpilib/wpilib/notifier.py]: I read that. And didn't comprehend it. And still don't. I'm kinda tired..

**Pull request merged**